chain-tx-enclave-next/mls/src/tree_math.rs

```diff
@@ -0,0 +1,334 @@
+    let lvl = level(n);
+    if lvl == 0 {
+        None
+    } else {
+        let x = lvl - 1;
+        // 01x -> 00x
+        Some(n ^ (0b01 << x))
+    }
+}
+
+/// The right child of an intermediate node.  Depends on the size of
+/// the tree because the straightforward calculation can take you
+/// beyond the edge of the tree.
+#[inline]
+pub fn right(x: NodeSize, leafs: LeafSize) -> Option<usize> {
+    let lvl = level(x);
+    if lvl == 0 {
+        None
+    } else {
+        let nodes = node_width(leafs);
+        // 01x -> 10x
+        let mut r = x ^ (0b11 << (lvl - 1));
+        while r >= nodes {
+            // TODO proof
+            r = left(r).unwrap();
+        }
+        Some(r)
+    }
+}
+
+/// The exponent of the largest power of 2 less than x. Equivalent to:
+///  int(math.floor(math.log(x, 2)))
+///
+/// return the index of the leading one bit, if all zero, return None.
+///
+/// ```
+/// use mls::tree_math::*;
+/// assert_eq!(log2(0), 0);
+/// assert_eq!(log2(1), 0);
+/// assert_eq!(log2(0b100001), 5);
+/// ```
+#[inline]
+pub fn log2(x: NodeSize) -> u32 {
+    #[cfg(feature = "nightly")]
+    let r = {
+        let msb = std::mem::size_of::<NodeSize>() as u32 * 8 - x.leading_zeros();
+        msb.saturating_sub(1)
+    };
+
+    #[cfg(not(feature = "nightly"))]
+    let r = {
+        if x == 0 {
+            return 0;
+        }
+
+        let mut k = 0;
+        while (x >> k) > 0 {
+            k += 1
+        }
+        k - 1
+    };
+
+    r
+}
+
+/// The index of the root node of a tree with n nodes
+/// return None for zero leafs.
+///
+/// ```
+/// use mls::tree_math::*;
+/// assert_eq!(root(0), 0);
+/// assert_eq!(root(1), 0);
+/// assert_eq!(root(0b10001), 0b011111);
+/// ```
+#[inline]
+pub fn root(leafs: LeafSize) -> NodeSize {
+    let nodes = node_width(leafs);
+    let lvl = log2(nodes);
+    ((1 as NodeSize) << lvl) - 1
+}
+
+/// The immediate parent of a node.  May be beyond the right edge of
+/// the tree.
+///
+/// Turning 00X or 10X to 01X
+/// where X is sequence of ones
+///
+/// ```
+/// use mls::tree_math::*;
+/// assert_eq!(parent_step(0b10011), 0b10111);
+/// assert_eq!(parent_step(0b11011), 0b10111);
+/// ```
+#[inline]
+pub fn parent_step(x: NodeSize) -> NodeSize {
+    let lvl = level(x);
+    let mask = ((x >> (lvl + 1)) & 0x01) << (lvl + 1);
+    (x | (1 << lvl)) ^ mask
+}
+
+/// The parent of a node.  As with the right child calculation, have
+/// to walk back until the parent is within the range of the tree.
+///
+/// ```
+/// use mls::tree_math::*;
+/// assert_eq!(parent(0b10001, 11), Some(0b10011));
+/// assert_eq!(parent(0b10001, 10), Some(0b01111));
+/// assert_eq!(parent(0b01011, 7), Some(0b00111));
+/// ```
+#[inline]
+pub fn parent(x: NodeSize, leafs: LeafSize) -> Option<NodeSize> {
+    let nodes = node_width(leafs);
+    assert!(x < nodes);
+    if x == root(leafs) {
+        return None;
+    }
+    let mut p = parent_step(x);
+    while p >= nodes {
+        p = parent_step(p)
```

Comment:
works without a semicolon at the end?

Reply:
It seems to work without it, but still I added the semicolon.